### PR TITLE
Add hugetlb-limit option to docker run/create

### PR DESCRIPTION
**Pull request by x1022as**
Signed-off-by: Deng Guangxing <dengguangxi

Please provide the following information:
-->

fixes https://github.com/moby/moby/issues/38113

**- What I did**

add hugetlb-limit option to docker run/create command to provide hugetlb limitation in container

**- How I did it**

add new options to create, and use cgroup to limit hugetlb usage

**- How to verify it**

```
root@5625cd1cf658:/go/src/github.com/docker/docker# docker run -ti --hugetlb-limit 2m:6m busybox cat /sys/fs/cgroup/hugetlb/hugetlb.2MB.limit_in_bytes
6291456
```

### CODE REVIEW

**Comment by justincormack** [comment]
The linter is complaining "opts/hugetlb.go:83:1: comment on exported method HugetlbOpt.GetAll should be of the form "GetAll ...""

**Comment by x1022as** [comment]
Thanks @justincormack , updated.

**Comment by thaJeztah** [comment]
We discussed this in the maintainers meeting and are okay with adding this, moving to code review

**Comment by dnephin** [comment]
Needs a rebase

**Comment by dnephin** [comment]
I noticed there is no issue open about this feature addition. It would be good to open an issue first to get an idea of how many users would actually make use of this feature.

**Comment by x1022as** [comment]
@dnephin updated.

I will open an issue about this.

File: docs/reference/run.md
Comment (justincormack) [inline docs/reference/run.md]: info -> into

File: man/docker-run.1.md
Comment (justincormack) [inline man/docker-run.1.md]: "Limit should be times of page size" -> "Limit should be a multiple of the huge page size"

File: man/docker-run.1.md
Comment (justincormack) [inline man/docker-run.1.md]: would -> will

File: pkg/sysinfo/sysinfo.go
Comment (justincormack) [inline pkg/sysinfo/sysinfo.go]: times -> multiple

File: pkg/sysinfo/sysinfo.go
Comment (justincormack) [inline pkg/sysinfo/sysinfo.go]: Do we already have a function for this?

File: pkg/sysinfo/sysinfo.go
Comment (justincormack) [inline pkg/sysinfo/sysinfo.go]: I don't think you should error here, just ignore this maybe the kernel has added some new filename.

File: pkg/sysinfo/sysinfo.go
Comment (x1022as) [inline pkg/sysinfo/sysinfo.go]: `HumanSize` in `go-units` takes 1KB = 1000B, and 1KiB as 1024B. users may not like the format like 4KiB, when they input 4k or 4kb, they mean 4096.

File: pkg/sysinfo/sysinfo.go
Comment (x1022as) [inline pkg/sysinfo/sysinfo.go]: if len(sres) is not 3,  huge page size may not lies in sres[1]. not sure if it's ok to ignore this length.

File: opts/hugetlb.go
Comment (dnephin) [inline opts/hugetlb.go]: `ss` is not a descriptive name. Could you give this variable a name that helps the reader understand what it contains?

File: opts/hugetlb.go
Comment (dnephin) [inline opts/hugetlb.go]: Why is this function necessary?

I think we should only support the formats that we support everywhere else from `units.RAMInBytes()`. Is there some reason we can't?

Looking at the code some more I'm guessing this is because the linux API excepts these values?

I'd like to propose this instead:
* parse both size using `units.RAMInBytes()` here in `opts` package
* have the API accept `int64` instead of string
* in `daemon.getHugetlbResources()` convert to the form expected by that API

I think that will remove all the duplicate validation, parsing and formatting. It will make the values consistent with other fields, and it will make the API consistent with newer API fields.

File: opts/hugetlb.go
Comment (dnephin) [inline opts/hugetlb.go]:
```go
errors.Wrapf(err, ...)
```

( using `github.com/pkg/errors`)

File: opts/hugetlb.go
Comment (x1022as) [inline opts/hugetlb.go]: @dnephin thanks for your advice. convert and format these values in `daemon.getHugetlbResources()` is a good idea. it will be convenient for users calling docker daemon with RestAPI.

However I think it's ok to keep `Size` as string type, because people using RestAPI may set Size as `2m`, `2M` ,  `2097152` ... ... Does this make sense to you,  any particular reason why we should make size int64 type as well?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: opts/hugetlb.go
Comment (dnephin) [inline opts/hugetlb.go]: Sure, that works, as long as we don't try and parse it to an int on the client.

**Review by vdemeester — COMMENTED** [review]
/cc @x1022as @thaJeztah @justincormack what is the status here ?

**Comment by x1022as** [comment]
I will rebase the code, ping @justincormack  and @thaJeztah

**Comment by thaJeztah** [comment]
Thanks @x1022as, give us a ping when you're done 👍

**Comment by x1022as** [comment]
updated, waiting for Jenkins

**Comment by cpuguy83** [comment]
```
21:19:39 # github.com/docker/docker/integration-cli
21:19:39 ./docker_cli_run_unix_test.go:503: undefined: dockerCmdWithStdoutStderr
21:19:39 ./docker_cli_run_unix_test.go:518: undefined: dockerCmdWithStdoutStderr
```

**Comment by x1022as** [comment]
Jenkins failed on  `Error response from daemon: No such container: docker-pr-z1289`

Is this known problem?

**Comment by cpuguy83** [comment]
This needs a rebase since the cli is no longer in this repo.

**Comment by x1022as** [comment]
integration test failed because of cli do not have hugetlb-limit flag yet.
add this flag to cli seem to need this patch in return.

any idea how to solve this? @cpuguy83

**Comment by cpuguy83** [comment]
You can do an API test instead.

File: docs/reference/commandline/create.md
Comment (cpuguy83) [inline docs/reference/commandline/create.md]: This should be removed

File: docs/reference/commandline/run.md
Comment (cpuguy83) [inline docs/reference/commandline/run.md]: This should be removed

File: docs/reference/run.md
Comment (cpuguy83) [inline docs/reference/run.md]: This should be removed

File: docs/reference/run.md
Comment (cpuguy83) [inline docs/reference/run.md]: This should be removed

File: integration-cli/docker_api_containers_unix_test.go
Comment (cpuguy83) [inline integration-cli/docker_api_containers_unix_test.go]: Need `testRequires(c, SameHostDaemon)` so the sysinfo check below works as expected.

Comment (cpuguy83) [inline integration-cli/docker_api_containers_unix_test.go]: Or query the value in a container.

File: man/docker-run.1.md
Comment (cpuguy83) [inline man/docker-run.1.md]: Remove

File: man/docker-run.1.md
Comment (cpuguy83) [inline man/docker-run.1.md]: Remove

File: opts/hugetlb.go
Comment (cpuguy83) [inline opts/hugetlb.go]: Needs to move to docker/cli repo (assuming we take this on the CLI).

File: opts/hugetlb.go
Comment (cpuguy83) [inline opts/hugetlb.go]: Why would the API accept the string format? This is a client concern. Sizes should generally be represented in the API as bytes.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: integration-cli/docker_api_containers_unix_test.go
Comment (cpuguy83) [inline integration-cli/docker_api_containers_unix_test.go]: It would be nice to deal with the actual struct here rather than the weird inspect output.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
One minor nit, but otherwise it looks good.
Can you rebase and push so that tests can re-run and make sure there's really no conflicts since it's been awhile since the last push.

**Comment by x1022as** [comment]
really sorry for late response, code updated and wait for janky.

**Comment by cpuguy83** [comment]
Some issues with power/z related to the kernel not supporting the new option.

**Comment by x1022as** [comment]
all green , @cpuguy83

File: pkg/sysinfo/sysinfo.go
Comment (cpuguy83) [inline pkg/sysinfo/sysinfo.go]: Why are we converting this to a string?

Comment (x1022as) [inline pkg/sysinfo/sysinfo.go]: for human readable information in inspect

Comment (thaJeztah) [inline pkg/sysinfo/sysinfo.go]: `docker inspect` should show the raw values; presenting values in a human-readable way is something to be handled in the CLI (e.g. when using `--format`)

Comment (x1022as) [inline pkg/sysinfo/sysinfo.go]: to use this directly in cgroup as well: 
```
# ls /sys/fs/cgroup/hugetlb/hugetlb.2MB.limit_in_bytes 
/sys/fs/cgroup/hugetlb/hugetlb.2MB.limit_in_bytes
```

Comment (cpuguy83) [inline pkg/sysinfo/sysinfo.go]: But docker should not be setting the cgroup directly.

Comment (x1022as) [inline pkg/sysinfo/sysinfo.go]: runtime-spec define hugetlblimit as below:
```
// LinuxHugepageLimit structure corresponds to limiting kernel hugepages
type LinuxHugepageLimit struct {
	// Pagesize is the hugepage size
	Pagesize string `json:"pageSize"`
	// Limit is the limit of "hugepagesize" hugetlb usage
	Limit uint64 `json:"limit"`
}
```

**Comment by AkihiroSuda** [comment]
ping @cpuguy83 LGTY?

File: pkg/sysinfo/sysinfo.go
Comment (cpuguy83) [inline pkg/sysinfo/sysinfo.go]: I would still like to get rid of this for consistency. All other string size values are using the units package, this should as well.

Comment (x1022as) [inline pkg/sysinfo/sysinfo.go]: one can get system default hugepage size in `2048 kB` format in `/proc/meminfo`, and need to be transformed into `2MB` format . go-unit take KB as decimal and kib as binary, this is the standard way, but people always take kB=1024B, so does cgroups :(

Comment (cpuguy83) [inline pkg/sysinfo/sysinfo.go]: If we need some specific conversion, then we should either do it in the units package , or directly where it's being read (sysinfo I guess here?).

However conversions from user input really should be consistent.

**Comment by mauri** [comment]
Any chance we can move this forward? @thaJeztah @cpuguy83

**Comment by cpuguy83** [comment]
I don't think my comment is addressed:

> If we need some specific conversion, then we should either do it in the units package , or directly where it's being read (sysinfo I guess here?).
>
> However conversions from user input really should be consistent.

**Comment by x1022as** [comment]
@cpuguy83  humanSize is implemented and called in sysinfo, to get and format system default hugepage size

for user input `--hugetlb-limit 2m:100m`, it's cli's job to parse and convert them, daemon take this struct as input:
```
type Hugetlb struct {
       PageSize string
       Limit    uint64
```

for user input, this is a problem indeed, because users use 100m as 100mib, go-units won't help, either. while as for this pr, I think it's ok.

not sure if I get your point :)

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
@justincormack PTAL

**Comment by justincormack** [comment]
Needs a rebase, but otherwise LGTM.

I think we may be over-validating and should leave more to `runc`...

**Comment by x1022as** [comment]
rebased， ping @justincormack  PTAL

File: daemon/daemon_unix.go
Comment (thaJeztah) [inline daemon/daemon_unix.go]: I don't think we do this for other flags, e.g. `--label a=b --label a=c` just overwrites the previous value, so not sure we need this.

File: pkg/sysinfo/sysinfo.go
Comment (thaJeztah) [inline pkg/sysinfo/sysinfo.go]: I see the proposed format for the CLI is `--hugetlb-limit 2m:6m` - in what case would `pageSize` be empty? (the flag format may need some discussion altogether)

File: pkg/sysinfo/sysinfo.go
Comment (thaJeztah) [inline pkg/sysinfo/sysinfo.go]: Do we actually need to know the default pagesize if no size is specified? reading the [documentation](https://github.com/torvalds/linux/blob/30955e71fc2bd1b966ce6531285ac59ae653819f/Documentation/vm/hugetlbpage.txt#L74-L76)

> Some platforms support multiple huge page sizes.  To allocate huge pages of a specific size, one must precede the huge pages boot command parameters with a huge page size selection parameter `"hugepagesz=<size>"`

i.e., if we want to use the default, do we have to specify anything?

File: pkg/sysinfo/sysinfo.go
Comment (thaJeztah) [inline pkg/sysinfo/sysinfo.go]: I have to agree with @justincormack that we may be doing too much validation. For example, we now not only validate if `size` is a valid size (i.e. it's in the list of configured sizes on the host), but also if we're able to parse/interpret that size to something useful.

Perhaps just validate:

- if the specified size is found on the machine (`isHugepageSizeValid()`) (or not even, but that can be discussed)
- if the specified limit is valid (i.e., we're able to parse the size, in bytes)

And leave the rest to the runtime/kernel.

File: pkg/sysinfo/sysinfo.go
Comment (thaJeztah) [inline pkg/sysinfo/sysinfo.go]: Do all these validation functions belong in `pkg/sysinfo`?  They seem to be related to the container configuration, not the system

File: pkg/sysinfo/sysinfo.go
Comment (thaJeztah) [reply pkg/sysinfo/sysinfo.go]: But if it's not 3, we can just skip, and see if the next file is in the expected format?

File: pkg/sysinfo/sysinfo.go
Comment (thaJeztah) [inline pkg/sysinfo/sysinfo.go]: looks like a leftover debug line?

File: pkg/sysinfo/sysinfo.go
Comment (thaJeztah) [inline pkg/sysinfo/sysinfo.go]: Better check if it's the suffix;

```go
if strings.HasSuffix(i.Name(), ".limit_in_bytes") {
```

File: pkg/sysinfo/sysinfo.go
Comment (thaJeztah) [inline pkg/sysinfo/sysinfo.go]: Wonder if we can make this less brittle; do we know if case is consistent, and if `k` / `K` / `kB` / `KB` are not mixed? For example, `1g` may be a valid size, but the pagesize on the host may be named `/sys/fs/cgroup/hugetlb/hugetlb.1GB.limit_in_byte`.


If we _do_ want to validate, perhaps;
- Pass sizes as _bytes_ through the API, not strings
- Get all the defined sizes from the host (all of `hugetlb.XX.limit_in_bytes`)
- Create a map to map their actual size in bytes (e.g. `9223372036854771712` for 1GB) to _name_ of the size (`1GB`)

If the OCI spec expects the size to be specified as a string, we can map bytes to the expected string if needed

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Left some comments; I'm not too familiar with this, but also saw something about ["transparent hugetbl" support](https://github.com/torvalds/linux/blob/30955e71fc2bd1b966ce6531285ac59ae653819f/Documentation/vm/transhuge.txt#L7-L10); 

> Transparent Hugepage Support is an alternative means of using huge pages for the backing of virtual memory with huge pages that supports the automatic promotion and demotion of page sizes and without the shortcomings of hugetlbfs.

If that's available on the host, does manual configuration still have advantages?

File: pkg/sysinfo/sysinfo.go
Comment (x1022as) [reply pkg/sysinfo/sysinfo.go]: one can have different page sizes and default size through configuration in system cmdline(default_hugepagesz=xxx hugepagesz=xxx .etc), it will be easier for sysinfo to handle this.

File: pkg/sysinfo/sysinfo.go
Comment (x1022as) [reply pkg/sysinfo/sysinfo.go]: I'd rather `--hugetlb-limit 6M` to be legal option, because `docker info` will display the default hugepage size(2MB .e.g), in such case `pagesize` is set to the default one, and 6M will be taken as limit.

File: pkg/sysinfo/sysinfo.go
Comment (x1022as) [reply pkg/sysinfo/sysinfo.go]: users do not have to specify, but we need to get default size to write `/sys/fs/cgroup/hugetlb/hugetlb.XXXX.limit_in_bytes` file

File: pkg/sysinfo/sysinfo.go
Comment (x1022as) [reply pkg/sysinfo/sysinfo.go]:
```
$echo 5242880 > hugetlb.2MB.limit_in_bytes
-bash: echo: write error: Invalid argument
```

I dont think this error message is kind of confused, so just validate limit here.

File: pkg/sysinfo/sysinfo.go
Comment (x1022as) [reply pkg/sysinfo/sysinfo.go]: I thought daemon host_config should take standard value as input, it's cli's job to parse user input into standard format.

**Comment by cpuguy83** [comment]
This seems replaced by #40160, closing.

**Pull request closed without merging**
